gateway-messages/src/sp_to_mgs.rs

```diff
@@ -590,6 +603,262 @@ impl fmt::Display for SpError {
             Self::SwitchDefaultImageError(code) => {
                 write!(f, "switch default image failed with code {code}")
             }
+            Self::Sprot(e) => write!(f, "sprot: {}", e),
+            Self::Spi(e) => write!(f, "spi: {}", e),
+            Self::Sprockets(e) => write!(f, "sprockets: {}", e),
+            Self::Update(e) => write!(f, "update: {}", e),
+            Self::Dump(e) => write!(f, "dump: {}", e),
+        }
+    }
+}
+
+// This is necessarily sparse for now. It's likely we'll clean up the sprockets
+// errors. These are ones that are capable of being reported by Sprot now.
+#[derive(
+    Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize, SerializedSize,
+)]
+pub enum SprocketsError {
+    BadEncoding,
+    UnsupportedVersion,
+
+    // When the type in hubris has been updated, but MGS does not yet know
+    // this type. The meaning of the error code here should be found in the
+    // `From<HubrisType> for MgsType` implementation in the hubris code.
+    Unknown(u32),
+}
+
+impl fmt::Display for SprocketsError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::BadEncoding => write!(f, "deserialization error"),
+            Self::UnsupportedVersion => write!(f, "unsupported version"),
+            Self::Unknown(code) => write!(f, "unknown error (code {})", code),
+        }
+    }
+}
+
+#[derive(
+    Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize, SerializedSize,
+)]
+pub enum DumperError {
+    SetupFailed,
+    UnalignedAddress,
+    StartReadFailed,
+    ReadFailed,
+    BadDumpAreaHeader,
+    WriteFailed,
+    HeaderReadFailed,
+    FailedToHalt,
+    FailedToResume,
+    FailedToResumeAfterFailure,
+    RegisterReadFailed,
+    ServerRestarted,
```

Comment:
Do we need to expose `ServerRestarted`? In general I think I'd expect these to be handled locally within the SP via retries, unless a task restarting means it panicked and lost some state that means it can't proceed. If we want to keep this one, we should probably rename it `TaskRestarted` to avoid confusion about what "server" means.

(Same comment on the other `ServerRestarted` variants later in this file.)

Reply:
I think we have to expose it. It's an error that can be returned from the API on non-idempotent operations. While it shouldn't happen N retries in a row, it might. I support changing the name though. Will do that.

Reply:
> Apologies for how tedious this request is, but to ensure we don't break these messages moving forward, could you:
> 
> 1. Bump [version::CURRENT](https://github.com/oxidecomputer/management-gateway-service/blob/caea9e57378c8d1989206c04d12cfe81f5b8aada/gateway-messages/src/lib.rs#L66) from 3 to 4?
> 2. Add a [v4](https://github.com/oxidecomputer/management-gateway-service/blob/caea9e57378c8d1989206c04d12cfe81f5b8aada/gateway-messages/tests/versioning/mod.rs#L9-L10) test module that checks the serialized forms of all the new errors?

Done in [060f6fa](https://github.com/oxidecomputer/management-gateway-service/pull/82/commits/060f6fa2371244cb1cd4d744e0ab3aa10d43f3a4)